_drafts/implementing-state-machines-in-postgresql.md

```diff
@@ -0,0 +1,301 @@
+elegant is a [user-defined aggregate](https://www.postgresql.org/docs/current/static/xaggr.html).
+
+In a nutshell, a user defined aggregate is a function that has an internal
+value (state) that is updated for each input value that is being passed into it
+using a state transition function. This means it fits our FSM model like a
+glove:
+
+```sql
+CREATE AGGREGATE order_events_fsm(text) (
+  SFUNC = order_events_transition,
+  STYPE = text,
+  INITCOND = 'start'
+);
+```
+
+The above defines a new aggregate called `order_events_fsm` which takes a
+`text` input (one of our events) and calls the `order_events_transition` state
+transition function (`FSUNC`) for each input along with the current state
+(`STYPE`) which is also of type `text`. The initial state is `start`
+(`INITCOND`).
+
+A quick test shows that it works as expected:
+
+```sql
+SELECT order_events_fsm(event ORDER BY id)
+FROM (VALUES
+  (1, 'create'),
+  (2, 'pay'),
+  (3, 'cancel')
+) examples(id, event);
+```
+```
+ order_events_fsm
+------------------
+ awaiting_refund
+```
+
+Now let's use our `order_events_fsm` to create a `BEFORE INSERT` trigger for
+our `order_events` table that makes sure all events of a given `order_id`
+are valid and don't lead to an error state. We do this using a simple `plpgsql`
+function that executes our `order_events_fsm` against all existing events
+of the current `order_id` plus the new event. If the final state is `error`,
+we raise an exception which causes the current transaction to roll back. It
+looks like this:
+
+
+```sql
+CREATE FUNCTION order_events_tigger_func() RETURNS trigger
+LANGUAGE plpgsql AS $$
+DECLARE
+  new_state text;
+BEGIN
+  SELECT order_events_fsm(event ORDER BY id)
+  FROM (
+    SELECT id, event FROM order_events WHERE order_id = new.order_id
+    UNION
+    SELECT new.id, new.event
+  ) s
+  INTO new_state;
+  
+  IF new_state = 'error' THEN
+    RAISE EXCEPTION 'invalid event';
+  END IF;
+
+  RETURN new;
+END
+$$;
+
+CREATE TRIGGER order_events_trigger BEFORE INSERT ON order_events
+FOR EACH ROW EXECUTE PROCEDURE order_events_tigger_func();
+```
+
+We can verify that this works by inserting a valid sequence of order events:
+
+```sql
+INSERT INTO order_events (order_id, event) VALUES
+  (1, 'create'),
+  (1, 'pay'),
+  (1, 'ship');
+```
+```
+INSERT 0 3
+```
+
+As well as an invalid sequence of events:
+
+```sql
+INSERT INTO order_events (order_id, event) VALUES
+  (2, 'create'),
+  (2, 'ship');
+```
+```
+psql:orders.sql:95: ERROR:  invalid event
+CONTEXT:  PL/pgSQL function order_events_tigger_func() line 14 at RAISE
+```
+
+As expected, only the first series of inserts made it into our table:
+
+```sql
+SELECT id, order_id, event FROM order_events;
+```
+```
+ id | order_id | event
+----+----------+--------
+  1 |        1 | create
+  2 |        1 | pay
+  3 |        1 | ship
+```
+
+If you're still on the fence about embedding this kind of logic into your db,
```

Comment:
Here it's "db", earlier it was "DB". Recommendation: write "database" everywhere :)